### PR TITLE
Source parser and stream utils

### PR DESCRIPTION
**Pull request by bishiboosh**
The main motivation for this PR is to replicate what's done in https://github.com/Kotlin/kotlinx.serialization/blob/master/formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt, because I'm using ktoml currently in an Android app and using streams directly without copying them to memory first would be a huge advantage to me.

The modifications done to the lib are the following:
- change `parseStringsToTomlTree` to use a sequence instead of a list.
- add a ktoml-source module, that parses directly from an okio `Source`
- use this module in `ktoml-file`, allowing to avoid copying the file to memory while parsing it
- adding JVM-only methods allowing to parse `InputStream`

I realize that this introduces JVM-only code, so I could spin the stream-related methods into a new project if needs be.

### CODE REVIEW

File: ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/parsers/TomlParser.kt
Comment (akuleshov7) [inline ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/parsers/TomlParser.kt:136]: Does it look less readable than the example with MutableList? :)

File: ktoml-source/src/commonMain/kotlin/com/akuleshov7/ktoml/source/TomlSourceReader.kt
Comment (akuleshov7) [inline ktoml-source/src/commonMain/kotlin/com/akuleshov7/ktoml/source/TomlSourceReader.kt:34]: `it` is fine here instead of `lines`

File: ktoml-source/src/commonMain/kotlin/com/akuleshov7/ktoml/source/TomlSourceReader.kt
Comment (akuleshov7) [inline ktoml-source/src/commonMain/kotlin/com/akuleshov7/ktoml/source/TomlSourceReader.kt:65]: same about `it`

File: ktoml-source/src/commonTest/kotlin/com/akuleshov7/ktoml/file/TomlSourceParserTest.kt
Comment (akuleshov7) [inline ktoml-source/src/commonTest/kotlin/com/akuleshov7/ktoml/file/TomlSourceParserTest.kt:53]: Do we still have `decodeFromFile`? 
Is it just an extension for the library or we are removing `decodeFromFile`?

File: ktoml-source/src/jvmMain/kotlin/com/akuleshov7/ktoml/source/JvmStreams.kt
Comment (akuleshov7) [inline ktoml-source/src/jvmMain/kotlin/com/akuleshov7/ktoml/source/JvmStreams.kt:12]: Do we really need JVM here? don't we want to have a common module for it?

File: ktoml-source/src/jvmTest/kotlin/com/akuleshov7/ktoml/source/StreamTests.kt
Comment (akuleshov7) [inline ktoml-source/src/jvmTest/kotlin/com/akuleshov7/ktoml/source/StreamTests.kt:176]: Better to have same syntax as we have for a straight forward decodeFromString:
`Toml.decodeFromString<Bool>("a = \"test\"")`

With a default object companion 

`@ThreadLocal
    public companion object Default : Toml(TomlConfig())`

**Review by akuleshov7 — APPROVED** [review]
Generally, looks good and should not affect a lot of people. But API changed, so we will need to take it at least into the `minor` release.

File: ktoml-source/src/commonMain/kotlin/com/akuleshov7/ktoml/source/SourceUtils.kt
Comment (akuleshov7) [inline ktoml-source/src/commonMain/kotlin/com/akuleshov7/ktoml/source/SourceUtils.kt:10]: is it possible to rename it to something more clear? :)
`readAndDecodeLines` or something like this?

**Comment by akuleshov7** [comment]
May be you can also add some small example to the readme? You have added new logic to the API, that can be useful for users. 

I fixed diktat/detekt code style issues and fixed some small typos

**Comment by bishiboosh** [comment]
@akuleshov7 I'll try to look at your comments today, but I may have to delay until next week, I'm a little bit swamped right now

**Comment by akuleshov7** [comment]
@bishiboosh 

You have been waiting for a review for a half of a year (sorry), so I do not think it is a problem for me to wait several weeks 😅

**Comment by akuleshov7** [comment]
Only today I have seen that this PR was opened in the same day when this awful war started one year ago.
A year has passed already and still it has not ended, oh :(

I have opened #188 to finalise this PR: we need also to think how to reduce the number of affected users of the library.

**Comment by akuleshov7** [comment]
And finally, after more than a year I have merged conflicts 🥳

File: buildSrc/src/main/kotlin/com/akuleshov7/Versions.kt
Comment (nulls) [inline buildSrc/src/main/kotlin/com/akuleshov7/Versions.kt]: 1.8.20?

Comment (akuleshov7) [reply buildSrc/src/main/kotlin/com/akuleshov7/Versions.kt]: good note, thanks, updated to 1.8.21

**Pull request merged**
